pmd-core/src/main/java/net/sourceforge/pmd/RuleBuilder.java

```diff
@@ -0,0 +1,246 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.commons.lang3.StringUtils;
+import org.w3c.dom.Element;
+
+import net.sourceforge.pmd.lang.Language;
+import net.sourceforge.pmd.lang.LanguageRegistry;
+import net.sourceforge.pmd.lang.LanguageVersion;
+import net.sourceforge.pmd.properties.PropertyDescriptor;
+
+
+/**
+ * Builds a rule, validating its parameters throughout. The builder can define property descriptors, but not override
+ * them. For that, use {@link RuleFactory#decorateRule(Rule, Element)}.
+ *
+ * @author Clément Fournier
+ * @since 6.0.0
+ */
+class RuleBuilder {
+
+    private List<PropertyDescriptor<?>> definedProperties = new ArrayList<>();
+    private String name;
+    private String clazz;
+    private Language language;
+    private String minimumVersion;
+    private String maximumVersion;
+    private String since;
+    private String message;
+    private String rulesetName;
+    private String externalInfoUrl;
+    private String description;
+    private List<String> examples = new ArrayList<>(1);
+    private RulePriority priority;
+    private boolean isDeprecated;
+    private boolean isUsesDfa;
+    private boolean isUsesMultifile;
+    private boolean isUsesTyperesolution;
+
+
+    RuleBuilder(String name, String clazz, String language) {
+        this.name = name;
+        language(language);
+        className(clazz);
+    }
+
+
+    public void usesDFA(boolean usesDFA) {
+        isUsesDfa = usesDFA;
+    }
+
+
+    public void usesMultifile(boolean usesMultifile) {
+        isUsesMultifile = usesMultifile;
+    }
+
+
+    public void usesTyperesolution(boolean usesTyperesolution) {
+        isUsesTyperesolution = usesTyperesolution;
+    }
+
+
+    private void language(String languageName) {
+        if (StringUtils.isBlank(languageName)) {
+            return;
+            // throw new IllegalArgumentException("Blank language attribute");
```

Comment:
why not throw?

Reply:
Java rules don't need the `language="java"` attribute, because [AbstractJavaRule](https://github.com/pmd/pmd/blob/0ff05ed74087282a516ffa3478ed52c42543664d/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/AbstractJavaRule.java#L20) calls `setLanguage` itself. So we have to wait until after the instantiation of the rule in the `build` method to know whether the rule has a language or not.

Many test cases in RuleSetFactoryTest use that behaviour and thus throwing an exception it would make them fail. Is that expected behaviour or should we get rid of it?

Reply:
Ok, let's leave it. Just delete this commented throw and let's put a comment in there for future self